### PR TITLE
TOOLS-23: static analysis

### PR DESCRIPTION
**Pull request by eugenetaranov-opsfleet**
staticanalysis: add the github CI config for static code analysis

This patch adds support for static code analysis with cppcheck inside the github CI.

Signed-off-by: Eugene Taranov <eugene@opsfleet.com>

### CODE REVIEW

**Comment by blogic** [comment]
could you fold the 5 patches into 1 single patch please ? also the patch is not properly annotated. 
OpenWrt offers a nice CheckList --> https://openwrt.org/submitting-patches#patch_checklist
The most important part is that the patch is missing the Signed-off-by line. 

The description should look something like this

staticanalysis: add the github setup yaml

This patch adds support for static code analyses inside the github CI.

Signed-off-by: $name <$email>

**Comment by blogic** [comment]
Is it possible to mask out folders int he scan or mark false positives ? The static analyzer picked this up
--> [feeds/wifi-trunk/hostapd/src/src/ap/ubus.c:989]: (error) Uninitialized variable: rem
34
which is a false positive

**Comment by eugenetaranov-opsfleet** [comment]
> Is it possible to mask out folders in the scan or mark false positives ? The static analyzer picked this up
> --> [feeds/wifi-trunk/hostapd/src/src/ap/ubus.c:989]: (error) Uninitialized variable: rem
> 34
> which is a false positive

@blogic, yes, I believe both ways are possible, will update

**Comment by blogic** [comment]
@eugenetaranov-opsfleet It might make sense to just whitelist selected folders as in the opensyc src/ folder

File: .github/workflows/static_analysis.yml [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+jobs:
+  static_analysis:
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v2
+    - name: install cppcheck
+      run: sudo apt install -y cppcheck
+    - name: Run cppcheck
+      run: cppcheck --error-exitcode=1 -j 2 feeds/wlan-ap/opensync/src
```
Comment (eugenetaranov-opsfleet) [inline .github/workflows/static_analysis.yml:17]: @blogic does it look good now?

File: .github/files/supressions_list.txt [hunk 2]
```diff
@@ -0,0 +1 @@
+memleak:feeds/wlan-ap/opensync/src/platform/openwrt/src/lib/target/src/uci_helper.c:372
```
Comment (blogic) [inline .github/files/supressions_list.txt:1]: This memaleak is actually present. The unitialized warning in the ubus plugin of hostapd was invalid

**Comment by eugenetaranov-opsfleet** [comment]
> could you fold the 5 patches into 1 single patch please ? also the patch is not properly annotated.
> OpenWrt offers a nice CheckList --> https://openwrt.org/submitting-patches#patch_checklist
> The most important part is that the patch is missing the Signed-off-by line.
> 
> The description should look something like this
> 
> staticanalysis: add the github setup yaml
> 
> This patch adds support for static code analyses inside the github CI.
> 
> Signed-off-by: $name <$email>

Could you please see description if it looks good to you @blogic ?
regarding `could you fold the 5 patches into 1 single patch please ?` github allows squashing commits during the merge https://github.blog/2016-04-01-squash-your-commits/, do you think its enough to solve or better to squash commits manually?

**Comment by blogic** [comment]
the description looks great, however it needs to go into the patch header, not the github PR.
you fold the patches easily by doing
git rebase -i 48eb65ab8e905cdab201bd67390707226156c08f~1
then use the r on the first patch to edit its description and f on all following ones to fold them

**Comment by blogic** [comment]
the checks just failed after the last commit, however i already fixed what cppcheck complains about and created a PR for it
([feeds/wlan-ap/opensync/src/platform/openwrt/src/lib/target/src/uci_helper.c:372]: (error) Memory leak: buffer
23) If it is ok for you i will just fold the patches while merging and add the description to the top patch.

**Comment by eugenetaranov-opsfleet** [comment]
> the checks just failed after the last commit, however i already fixed what cppcheck complains about and created a PR for it
> ([feeds/wlan-ap/opensync/src/platform/openwrt/src/lib/target/src/uci_helper.c:372]: (error) Memory leak: buffer
> 23) If it is ok for you i will just fold the patches while merging and add the description to the top patch.

I think I did it, take a look now please @blogic

**Comment by blogic** [comment]
merged into trunk, Thanks !

**Pull request closed without merging**